src/main/java/org/datadog/jenkins/plugins/datadog/traces/write/TraceWriter.java

```diff
@@ -0,0 +1,118 @@
+package org.datadog.jenkins.plugins.datadog.traces.write;
+
+import hudson.model.Run;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.logging.Logger;
+import net.sf.json.JSONObject;
+import org.datadog.jenkins.plugins.datadog.DatadogClient;
+import org.datadog.jenkins.plugins.datadog.DatadogUtilities;
+import org.datadog.jenkins.plugins.datadog.model.BuildData;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+
+public class TraceWriter {
+
+    private static final Logger logger = Logger.getLogger(TraceWriter.class.getName());
+
+    private static final String QUEUE_CAPACITY_ENV_VAR = "DD_JENKINS_TRACES_QUEUE_CAPACITY";
+    private static final String SUBMIT_TIMEOUT_ENV_VAR = "DD_JENKINS_TRACES_SUBMIT_TIMEOUT_SECONDS";
+    private static final String STOP_TIMEOUT_ENV_VAR = "DD_JENKINS_TRACES_STOP_TIMEOUT_SECONDS";
+    private static final String POLLING_INTERVAL_ENV_VAR = "DD_JENKINS_TRACES_POLLING_INTERVAL_SECONDS";
+    private static final String BATCH_SIZE_LIMIT_ENV_VAR = "DD_JENKINS_TRACES_BATCH_SIZE_LIMIT";
+    private static final int DEFAULT_QUEUE_CAPACITY = 10_000;
+    private static final int DEFAULT_SUBMIT_TIMEOUT_SECONDS = 30;
+    private static final int DEFAULT_STOP_TIMEOUT_SECONDS = 15;
+    private static final int DEFAULT_POLLING_INTERVAL_SECONDS = 10;
+    private static final int DEFAULT_BATCH_SIZE_LIMIT = 100;
+
+    private final TraceWriteStrategy traceWriteStrategy;
+    private final BlockingQueue<JSONObject> queue;
+    private final Thread poller;
+
+    public TraceWriter(DatadogClient datadogClient) {
+        this.traceWriteStrategy = datadogClient.createTraceWriteStrategy();
+
+        this.queue = new ArrayBlockingQueue<>(getEnv(QUEUE_CAPACITY_ENV_VAR, DEFAULT_QUEUE_CAPACITY));
+
+        this.poller = new Thread(new QueuePoller(traceWriteStrategy, queue), "DD-Trace-Writer");
+        this.poller.setDaemon(true);
+        this.poller.start();
+    }
+
+    public void submitBuild(final BuildData buildData, final Run<?,?> run) throws InterruptedException, TimeoutException {
+        JSONObject buildJson = traceWriteStrategy.serialize(buildData, run);
+        submit(buildJson);
+    }
+
+    public void submitPipeline(FlowNode flowNode, Run<?, ?> run) throws InterruptedException, TimeoutException {
+        Collection<JSONObject> nodeJsons = traceWriteStrategy.serialize(flowNode, run);
+        for (JSONObject nodeJson : nodeJsons) {
+            submit(nodeJson);
+        }
+    }
+
+    private void submit(JSONObject json) throws InterruptedException, TimeoutException {
+        if (!queue.offer(json, getEnv(SUBMIT_TIMEOUT_ENV_VAR, DEFAULT_SUBMIT_TIMEOUT_SECONDS), TimeUnit.SECONDS)) {
```

Comment:
Should we check if the writer has stopped before `queue.offer`? This would prevent to add more items to the queue in case the thread must stop.

Reply:
I think it makes more sense to allow the caller to add the span here: if we reject it, the span is not submitted. If we allow it to be added, then it will almost certainly manage to get flushed (since the poller thread waits for a few seconds before shutting down, which should be enough time to call submit for everyone who retains reference to the old writer). And if it doesn't get flushed, nothing bad will happen: it'll be GCed together with the queue.